### PR TITLE
Fix missing transaction_index in contract result by hash response

### LINKED ISSUE #4893: Contract Results Details API omits transaction_index on failed transaction
### Description

The relay relies on the results from https://testnet.mirrornode.hedera.com/api/v1/contracts/results/<transactionHash> to provide valid transaction receipts.
However, it seems when there's a failed Ethereum transactions the mirror node omits the transaction index e.g. https://testnet.mirrornode.hedera.com/api/v1/contracts/results/0x36fd750e33f0440eef56444dd2afdf0c96ddcd0cfdf461faeebdd5282033819c

This results in a null value in the relay therefore returning a null transaction index to consumers of the relay which is inaccurate. A valid transaction index should be populated even on failure.

See also https://github.com/hashgraph/hedera-json-rpc-relay/issues/1151

### Steps to reproduce

1. Find a failed ethereum transaction
2. Retrieve the transaction hash
3. Call `https://testnet.mirrornode.hedera.com/api/v1/contracts/results/<transactionHash>` e.g. https://testnet.mirrornode.hedera.com/api/v1/contracts/results/0x0339477ac4e9516be0bca4e1634b250c8fe431b4fce1ced136eccb516ed456fa

### Additional context

Testnet has been reset so above URL no longer works. Here's the output from when it did:
```json
{
   "address":null,
   "amount":0,
   "bloom":"0x...",
   "call_result":"0x",
   "contract_id":null,
   "created_contract_ids":[
      
   ],
   "error_message":"INVALID_ACCOUNT_ID",
   "from":"0x0000000000000000000000000000000000000386",
   "function_parameters":"0x...",
   "gas_limit":0,
   "gas_used":0,
   "timestamp":"1667866727.240345165",
   "to":null,
   "hash":"0x",
   "block_hash":"0xc9a5bd67f1cfe4007774558a6368402b9bed72a113573bcf58b1feb065aa2c26c3413616a6cc729e2a6bb453dd2e6f3e",
   "block_number":26925964,
   "logs":[
      
   ],
   "result":"UNKNOWN",
   "state_changes":[
      
   ],
   "status":"0x0",
   "failed_initcode":"0x...",
   "access_list":"0x",
   "block_gas_used":0,
   "chain_id":"0x128",
   "gas_price":"0x",
   "max_fee_per_gas":"0x0ae",
   "max_priority_fee_per_gas":"0x0",
   "r":"0xdbf6c25927513afc9b23893b63318da93eeb60d7532dc8da9b0eeb1e3fc87f3a",
   "s":"0x62575b5644e97cfe5f8a9621e8e97f5d0724089ab032171b01c9fe918a56f8",
   "type":2,
   "v":0,
   "nonce":15
}
```

### Hedera network

testnet

### Version

v0.67.3

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by xin-hedera**
**Description**:

This PR fixes the missing transaction index in contract response bug

- Persist default contract result with sensible value for certain failed ethereum transactions
- Add a db migration to back fill failed ethereum transaction contract results
- Remove the default contract result for ethereum transaction logic in REST API
- Return 404 and log error if no contract result found for an ethereum transaction

**Related issue(s)**:

Fixes #4893 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java [hunk 1]
```diff
@@ -80,7 +84,17 @@ public void process(@NonNull RecordItem recordItem, Transaction transaction) {
         // handle non create/call transactions
         var contractCallOrCreate = isContractCreateOrCall(transactionBody);
         if (!contractCallOrCreate && !isValidContractFunctionResult(functionResult)) {
-            // if transaction is neither a create/call and has no valid ContractFunctionResult then skip
+            var status = recordItem.getTransactionRecord().getReceipt().getStatus();
+            if (!recordItem.isSuccessful()
+                    && status != ResponseCodeEnum.DUPLICATE_TRANSACTION
+                    && status != ResponseCodeEnum.WRONG_NONCE
+                    && transactionBody.hasEthereumTransaction()) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java:91]: Are we sure that contract call/create can't also be missing a contract result?

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/BackfillFailedEthereumTransactionContractResultMigrationTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+class BackfillFailedEthereumTransactionContractResultMigrationTest extends IntegrationTest {
+
+    private static final byte[] EMPTY = new byte[] {};
+
+    private final ContractResultRepository contractResultRepository;
+
+    private final JdbcTemplate jdbcTemplate;
+
+    @Value("classpath:db/migration/v1/V1.79.1__backfill_ethereum_transaction_contract_result.sql")
+    private final File sql;
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/BackfillFailedEthereumTransactionContractResultMigrationTest.java:52]: Prefer Spring's `Resource` and treat it as an InputStream.
```suggestion
    private final Resource sql;
```

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/BackfillFailedEthereumTransactionContractResultMigrationTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.test.context.TestPropertySource;
+
+@EnabledIfV1
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Tag("migration")
+@TestPropertySource(properties = "spring.flyway.target=1.79.0")
+class BackfillFailedEthereumTransactionContractResultMigrationTest extends IntegrationTest {
+
+    private static final byte[] EMPTY = new byte[] {};
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/BackfillFailedEthereumTransactionContractResultMigrationTest.java:45]: Use `DomainUtils.EMPTY_BYTE_ARRAY`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java [hunk 4]
```diff
@@ -56,6 +58,8 @@
 @RequiredArgsConstructor
 public class ContractResultServiceImpl implements ContractResultService {
 
+    private static final byte[] EMPTY = new byte[0];
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java:61]: Use `DomainUtils.EMPTY_BYTE_ARRAY`

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.79.1__backfill_ethereum_transaction_contract_result.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+-- backfill default contract result for failed ethereum transactions
+insert into contract_result (call_result, consensus_timestamp, contract_id, function_parameters, gas_limit, gas_used, payer_account_id, transaction_hash, transaction_index, transaction_nonce, transaction_result)
+select '\x', et.consensus_timestamp, 0, coalesce(et.call_data, '\x'), et.gas_limit, 0, et.payer_account_id, et.hash, t.index, t.nonce, t.result
+from ethereum_transaction et
+join transaction t on t.consensus_timestamp = et.consensus_timestamp
+left join contract_result cr on cr.consensus_timestamp = et.consensus_timestamp
+where cr.consensus_timestamp is null and t.result not in (11, 22, 104, 220, 312);
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.79.1__backfill_ethereum_transaction_contract_result.sql:7]: 22, 104, 220 are the three response codes we check in RecordItem.checkSuccess

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java (hunk 1 above)
Comment (xin-hedera) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java]: In `ContractResultServiceImpl`, a minimal `ContractResult` is created for contract call / create without contract result in transaction record.

there are 9 such contract call / create TXs without contract result in db for mainnet, that's before we had the above minimal `ContractResult` logic in place

**Review by steven-sheehy — APPROVED** [review]

**Review by edwin-greene — APPROVED** [review]
Looks good to me.

**Pull request merged**